### PR TITLE
RFC: Proposal for Cluster Mesh

### LINKED ISSUE #12009: Clustermesh implementation proposal + CRD
Write implementation proposal for the clustermesh feature (https://github.com/kubermatic/kubermatic/issues/11521) + add clustermesh CRDs.

### PR DESCRIPTION
**Pull request by rastislavs**
**What this PR does / why we need it**:

Adds proposal for the Cluster Mesh feature (https://github.com/kubermatic/kubermatic/issues/11521)

**Which issue(s) this PR fixes**:

Fixes #12009 

**What type of PR is this?**
/kind documentation
/kind design

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
NONE
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

File: docs/proposals/cluster-mesh.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  // APIServerServiceType defines the k8s service type used to expose the clustermesh apiserver the other clusters.
+  APIServerServiceType v1.ServiceType `json:"apiServerServiceType,omitempty"`
+
+  // APIServerIPs contains a list of IPs on which the clustermesh apiserver of this cluster is exposed to the other clusters.
+  // It can be a single IP if APIServerServiceType == LoadBalancer, or multiple IPs if APIServerServiceType == NodePort.
+  APIServerIPs []string `json:"apiServerIPs,omitempty"`
+
+  // APIServerPort defines the port on which the clustermesh apiserver of this cluster is exposed to the other clusters.
+  APIServerPort int `json:"apiServerPort,omitempty"`
+}
```
Comment (embik) [inline docs/proposals/cluster-mesh.md]: how would I reference a KKP user cluster from this struct? As far as I understood the other documentation, you create one `ClusterMesh` resource for X clusters you want to interconnect, but I don't see a way to actually list the X clusters I want.

File: docs/proposals/cluster-mesh.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+```
+
+A new cluster-mesh controller on the seed level  (EE only) will make sure that:
+
+- For each `ClusterMesh`, cluster mesh CA certificates are issued and stored in a secret in the kubermatic namespace,
+- Each cluster in the `ClusterMesh` list is assigned with an unique cluster ID (1 byte) and stored in the `ClusterMesh` CR,
```
Comment (embik) [inline docs/proposals/cluster-mesh.md:92]: I don't think the `cluster-mesh-controller` should alter the `ClusterMesh` spec. Maybe we should have a status field for information the controller produces and needs to remember?

File: docs/proposals/cluster-mesh.md (hunk 1 above)
Comment (rastislavs) [reply docs/proposals/cluster-mesh.md]: Damn, OFC, I completely forgot to include the KKP cluster name in the struct

File: docs/proposals/cluster-mesh.md (hunk 2 above)
Comment (rastislavs) [reply docs/proposals/cluster-mesh.md:92]: Fair point. I originally thought abut having the ability to optionally define some stuff like clusterMesh ID by the user, but its probably better to make an exact distinction on what needs to be defined by the user / vs assigned by controller (via status). Will update that.

**Pull request merged**